### PR TITLE
Add installation instructions

### PR DESCRIPTION
**Pull request by grunweg**
I tried my hand at writing some instructions, which could go into the readme.

While at it, I also made two minor changes to your readme. Feel free to accept or decline those as you prefer!

### CODE REVIEW

**Comment by grunweg** [comment]
Here are the proposed installation instructions.
edit: pushed them as a separate commit now --- figuring out how to do this wasn't hard after all :-)

XXX what's the minimum required rust version?

**Comment by dzfranklin** [comment]
Thanks! I've been meaning to write something like that for a while. I like your phrasing.

I'll probably hold off on publishing an update until I've finished the changes I'm working on around better supporting tests in main.

Edit: I've got no idea what the minimum Rust version is, I guess I'll have fun with rustup

File: README.md [hunk 1]
```diff
@@ -12,8 +12,10 @@ A light wrapper around rr, the time-travelling debugger.
 > changed their value or when a function was last called on an object (using
 > conditional breakpoints). [(source)][about-quote-source]
 
-Suppose we ran a test and got a failure. We first re-run the test under `rr` to
-record the entire execution (including everything else on your system the test
+
+## Example demonstration
```
Comment (dzfranklin) [inline README.md]:
```suggestion
## Example
```

File: README.md [hunk 2]
```diff
@@ -74,5 +76,20 @@ Note that since we're going in reverse the old and new values are backwards.
 `-1992923200` is whatever used to be at the address `0x000055dcac485ed9`
 before it was used for `a`.
 
+## Installation
+
+`cargo-rr` is a [custom cargo command](custom-cargo-commands). It can be installed
```
Comment (dzfranklin) [inline README.md]:
```suggestion
`cargo-rr` is a [custom cargo subcommand](custom-cargo-subcommands). It can be installed
```

File: README.md [hunk 3]
```diff
@@ -74,5 +76,20 @@ Note that since we're going in reverse the old and new values are backwards.
...
+in a terminal.This will download the source code for cargo-rr and all its dependencies and compile the package from source.
+
+Upon success, you can run `cargo rr` in the terminal to access `cargo-rr`.
+Usually, you want to run onen of `cargo-rr`'s subcommands (such as `test` for recording the execution of a test, `run` for recording the execution of a binary or `replay` to replay an ``rr`` trace you recorded using one of the previous commands).
+Running `cargo rr help` or just `cargo rr` will display a list of all available subcommands.
+
+
 [about-quote-source]: https://developer.chrome.com/blog/chromium-chronicle-13/
 [crates-io]: https://crates.io/crates/cargo-rr
+[custom-cargo-commands]: https://doc.rust-lang.org/stable/book/ch14-05-extending-cargo.html
```
Comment (dzfranklin) [inline README.md]:
```suggestion
[custom-cargo-subcommands]: https://doc.rust-lang.org/stable/book/ch14-05-extending-cargo.html
```

**Comment by grunweg** [comment]
Sure, I don't think this is very urgent.
About the rust version: I successfully installed cargo-rr using the 2021-04-24 nightly; 1.51.0 stable also works. (Those were the toolchains I had lying around.) I'll leave further fun to you :-) I know Debian includes rustc 1.34 (on armel) resp. 1.41.0 (everywhere else); those might be interesting to test.

File: README.md [hunk 4]
```diff
@@ -74,5 +76,20 @@ Note that since we're going in reverse the old and new values are backwards.
...
+## Installation
+
+`cargo-rr` is a [custom cargo command](custom-cargo-commands). It can be installed
+by installing the `cargo-rr` package from `crates.io`: just run
+```bash
+cargo install cargo-rr
+```
+in a terminal.This will download the source code for cargo-rr and all its dependencies and compile the package from source.
+
+Upon success, you can run `cargo rr` in the terminal to access `cargo-rr`.
```
Comment (dzfranklin) [inline README.md]:
```suggestion
in a terminal. After installing you can run `cargo rr` in the terminal to access `cargo-rr`.
```

File: README.md [hunk 5]
```diff
@@ -74,5 +76,20 @@ Note that since we're going in reverse the old and new values are backwards.
...
+Usually, you want to run onen of `cargo-rr`'s subcommands (such as `test` for recording the execution of a test, `run` for recording the execution of a binary or `replay` to replay an ``rr`` trace you recorded using one of the previous commands).
+Running `cargo rr help` or just `cargo rr` will display a list of all available subcommands.
+
+
```
Comment (dzfranklin) [inline README.md]: I'd like to expand this into its own section.

```suggestion
## Usage

Run `cargo rr test` or `cargo rr run` with any options you'd normally give to `cargo test` or `cargo run`. For example, you might run `cargo rr test --test my_integration_test some_filter`.

Once you've made a recording you can replay the last recording in a debugger with `cargo rr replay`.

### Advanced

You can pass options to rr by putting them after the separater `--`. If you do so, you are responsible for making sure the options you set don't conflict with the options `cargo-rr` sets for you. In general, you should be fine if you avoid customizing where traces go or what they're named. Please file a bug if you run into a conflict so I can add the option you want to `cargo-rr`.

`cargo rr test` and `cargo rr run` call `rr record` under the hood, so you can see the full list of rr options by running `rr record -h`. For example, `cargo rr test --test my_test -- --chaos` runs the tests under chaos mode, which randomizes scheduling decisions to try to surface concurrency bugs.

`cargo rr replay` calls `rr replay` under the hood, so you can see the full list of rr options by running `rr replay -h`. For example, `rr replay -- --stats=10000` displays brief stats every 10,000 steps.

Run `cargo rr help` to see the full usage.
```

**Review by dzfranklin — CHANGES_REQUESTED** [review]
Overall, I like this. I've made a few minor nitpicky changes, and expanded the usage section to include some things I'd meant to explain.

**Comment by grunweg** [comment]
Thanks for the feedback! I like those suggestions and just added them.

One more idea: as I understand it, `cargo-rr` passes the `rust-gdb` script to rr automatically (via the `--debugger` option). I was looking for that a bit today; mentioning this could also help :-) Perhaps like the following:

> `cargo-rr` configures `rr` to use the rust pretty-printer for `gdb` (by passing the `-d rust-gdb` option to `rr`).

**Comment by dzfranklin** [comment]
> cargo-rr configures rr to use the rust pretty-printer for gdb (by passing the -d rust-gdb option to rr).

I like it. What do you think about putting it in the advanced section, and phrasing it something like

> cargo-rr configures rr to use the rust pretty-printer for gdb, so you don't need to (by passing the option -d rust-gdb to rr).

**Comment by grunweg** [comment]
Sounds good to me! Feel free to add this to this branch directly.

**Pull request merged**